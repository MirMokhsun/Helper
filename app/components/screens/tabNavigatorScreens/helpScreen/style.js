import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#F5F5F7',
    },
    wrapperToogle: {
        width: '100%',
        paddingHorizontal: 16,
        paddingVertical: 8
    },
    containerTraidingApp: {
        width: 'auto',
        minHeight: 66,
        borderBottomWidth: 1,
        borderBottomColor: '#D3D9E7',
        marginHorizontal: 15,
        flexDirection: 'row',
        alignItems: 'center'
    },
    buttonTraidingApp: {
        fontFamily: 'Roboto-Light',
        fontSize: 16,
        color: '#000756',
        lineHeight: 21,
        flex: 1,
        paddingTop: 15,
        paddingBottom: 15,
    },
    containerToogle: {
        width: '100%',
        height: 32,
        backgroundColor: '#EBECEE',
        borderRadius: 16,
        padding: 4
    },
    wrapperTextInToogle: {
        position: 'absolute',
        top: 4,
        left: 4,
        width: '100%',
        height: '100%',
        flexDirection: 'row',
        zIndex: 2
    },
    textInToogle: {
        flex: 1,
        textAlignVertical: 'center',
        textAlign: 'center',
        fontSize: 16
    },
    sliderInToogle: {
        width: '50%',
        height: 24,
        borderRadius: 12,
        backgroundColor: '#FFFFFF'
    },
    activeTab: {
        fontFamily: 'Roboto-Bold',
        color: "#000756"
    },
    notActiveTab: {
        fontFamily: 'Roboto-Light',
        color: "#9FA0B2"
    },
});