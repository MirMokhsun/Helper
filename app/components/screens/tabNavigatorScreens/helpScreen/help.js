/* eslint-disable react/no-array-index-key */
/* eslint-disable no-plusplus */
/* eslint-disable no-unused-expressions */
import React, { Component } from 'react';
import { View, FlatList, Text, Animated, TouchableWithoutFeedback, Easing, I18nManager, TouchableOpacity, Platform, Linking } from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { ListPicker } from '../../../commonComponents/listPicker/listPicker';
import { styles } from './style';
import { localization } from '../../../../common/localization';
import Help from '../../../commonComponents/svg/Help';
import Utils from '../../../../api/utils/utils';
import { withNavigation } from 'react-navigation';
import { dictionary } from '../../../../common/faq/dictionary';
import { faq } from '../../../../common/faq/faq';
import { isUndefinedReturnSecondParam } from '../../../../api/isUndefinedReturnSecondParam/isundefinedreturnsecondparam';
import { getUrl } from '../../../../api/getUrlForLinkTraidingApp/getUrlForLinkTraidingApp';

const { size, setTextAlign, isIOS } = Utils;
const maxLeftDistance = size.width / 2 - 20;
const animationDuration = 100;

export class HelpScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    tabBarLabel: isUndefinedReturnSecondParam(navigation.state.params, 'pageTitle', 'Help'),
    tabBarIcon: ({ focused }) => {
      return focused ? <Help color="#1459D2" /> : <Help color="#000756" />;
    },
  })

  constructor(props) {
    super(props);
    this.maxLeftDistance = I18nManager.isRTL ? -maxLeftDistance : maxLeftDistance;
    this.state = {
      leftDistance: new Animated.Value(0),
      isFaq: true,
      styleFaqText: styles.activeTab,
      styleVocabularyText: styles.notActiveTab,
      dictionaryFlatList: null,
    }
    this.couner = 0;
  }

  componentWillMount() {
    const { isFaq } = this.state;
    const { language, navigation, filterHelpText } = this.props;
    navigation.setParams({ pageTitle: localization[language].Help });
    this.setDictionaryFlatList(isFaq, language, filterHelpText);
  }

  componentDidUpdate(prevProps, prevState) {
    const { isFaq } = this.state;
    const { language, filterHelpText, navigation } = this.props;
    if (prevState.isFaq !== isFaq || prevProps.language !== language || prevProps.filterHelpText !== filterHelpText) {
      this.setDictionaryFlatList(isFaq, language, filterHelpText);
    }
    if (language !== prevProps.language) {
      navigation.setParams({ pageTitle: localization[language].Help });
    }
  }

  animation = (_toDistance, _fromDistane) => {
    const { leftDistance } = this.state;
    Animated.timing(
      leftDistance,
      {
        fromValue: _fromDistane,
        toValue: _toDistance,
        duration: animationDuration,
        easing: Easing.linear,
        useNativeDriver: true
      }
    ).start();
  }

  toogleHelp = () => {
    const { isFaq } = this.state;
    let styleFaqText;
    let styleVocabularyText;
    if (isFaq) {
      styleFaqText = styles.notActiveTab;
      styleVocabularyText = styles.activeTab;
      this.animation(this.maxLeftDistance, 0)
    } else {
      styleFaqText = styles.activeTab;
      styleVocabularyText = styles.notActiveTab;
      this.animation(0, this.maxLeftDistance);
    }
    this.setState({
      styleFaqText,
      styleVocabularyText,
      isFaq: !isFaq
    });
  }

  createArrayListPickers = (language, arr, filterText) => {
    const { navigation } = this.props;
    const currentLanguage = localization[language];
    const filter = filterText ? filterText.toLowerCase() : '';
    const { isFaq } = this.state;
    const arrWithData = arr[language];
    const arrWithItems = [];
    const currentTextAlign = setTextAlign(language);

    arrWithData.forEach((item, i) => {
      if (arrWithData[i].title.toLowerCase().includes(filter)) {
        arrWithItems.push((<ListPicker
          id={`DetailArticleScreen_${i}`}
          text={arrWithData[i].title}
          key={i}
          currentTextAlign={currentTextAlign}
          onPress={() => {
            navigation.navigate("DetailArticleScreen", { index: i, isFaq, language, title: currentLanguage.Help } );
          }}
        />))
      }
    });
    if (filterText === '') {
      arrWithItems.unshift(
        <TouchableOpacity style={styles.containerTraidingApp} key={9999999999} onPress={this.goToTraidingApp}>
          <Text style={[styles.buttonTraidingApp, { textAlign: currentTextAlign }]}>
            {currentLanguage.OurApps}
          </Text>
        </TouchableOpacity>
      )
    }
    return arrWithItems;
  }

  goToTraidingApp = () => {
    const { language } = this.props;
    const url = getUrl(language, isIOS);
    Linking.openURL(url);
  }

  setDictionaryFlatList = (isFaq, language, filterHelpText) => {
    let result = faq;
    if (!isFaq) {
      result = dictionary
    }
    this.setState({ dictionaryFlatList: this.createArrayListPickers(language, result, filterHelpText) })
  }

  render() {
    const { styleFaqText, styleVocabularyText, leftDistance, dictionaryFlatList } = this.state;
    const { language } = this.props;
    const currentLanguage = localization[language];
    return (
      <View testID='welcome' style={styles.container}>
        <View style={styles.wrapperToogle}>
          <TouchableWithoutFeedback onPress={this.toogleHelp} testID="HelpScreenBtnID" accessibilityLabel="HelpScreenBtnID">
            <View style={styles.containerToogle} >
              <View style={styles.wrapperTextInToogle}>
                <Text style={[styles.textInToogle, styleFaqText]}>FAQ</Text>
                <Text style={[styles.textInToogle, styleVocabularyText]}>{currentLanguage.Glossary}</Text>
              </View>
              <Animated.View style={[styles.sliderInToogle, { transform: [{ translateX: leftDistance }] }]} ></Animated.View>
            </View>
          </TouchableWithoutFeedback>
        </View>
        <FlatList data={dictionaryFlatList} renderItem={({ item }) => item} />
      </View>
    );
  }
}

const mapStateToProps = state => ({
  language: state.currentLanguage.language,
  filterHelpText: state.searchInHelp.filterHelpText,
})

const mapDispatchToProps = dispatch => bindActionCreators({
}, dispatch)

export default withNavigation(connect(mapStateToProps, mapDispatchToProps)(HelpScreen));