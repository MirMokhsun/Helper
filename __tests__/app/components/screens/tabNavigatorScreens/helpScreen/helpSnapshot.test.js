import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';
import { HelpScreen } from '../../../../../../app/components/screens/tabNavigatorScreens/helpScreen/help';

const mockPlatform = OS => {
    jest.resetModules();
    jest.doMock("Platform", () => ({ OS, select: objs => objs[OS] }));
};
const mockLinking = () => {
    jest.mock('Linking', () => {
        return {
            openURL: jest.fn(() => Promise.resolve('success'))
        }
    })
}

describe('Help', () => {
    test('create snapshot HelpScreen', () => {
        const renderer = new ShallowRenderer();
        let helpScreen = renderer.render(<HelpScreen language={"ru"} navigation={{ setParams: () => { } }} />);
        expect(helpScreen).toMatchSnapshot();
    });

    describe('createArrayListPickers', () => {
        test('test function createArrayListPickers ListPicker length', () => {
            let helpScreen = renderer.create(<HelpScreen language={"ru"} navigation={{ setParams: () => { } }} />).getInstance();
            expect(helpScreen.createArrayListPickers(
                'ru',
                { ru: [{ title: '', text: '' }, { title: '', text: '' }, { title: '', text: '' }, { title: '', text: '' }] },
                '').length
            ).toEqual(5);
        });

        test('test function createArrayListPickers ListPicker item value', () => {
            let helpScreen = renderer.create(<HelpScreen language={"ru"} navigation={{ setParams: () => { } }} />).getInstance();
            let getResult = helpScreen.createArrayListPickers(
                'ru',
                { ru: [{ title: 'test1', text: '' }, { title: 'test2', text: '' }, { title: 'test3', text: '' }, { title: 'test4', text: '' }] },
                '');
            let getRendered = renderer.create(getResult[2]).getInstance();
            expect(getRendered.props.text).toEqual('test2');
        });

        test('test function createArrayListPickers ListPicker filterText', () => {
            let helpScreen = renderer.create(<HelpScreen language={"en"} navigation={{ setParams: () => { } }} />).getInstance();
            let getResult = helpScreen.createArrayListPickers(
                'ru',
                { ru: [{ title: 'test1', text: '' }, { title: 'hello', text: '' }, { title: 'test3', text: '' }, { title: 'test4', text: '' }] },
                'hello');
            let getRendered = renderer.create(getResult[0]).getInstance();
            expect(getRendered.props.text).toEqual('hello');
        });
    });

    test('function toogleHelp state isFaq', () => {
        let helpScreen = renderer.create(<HelpScreen language={"ru"} navigation={{ setParams: () => { } }} />).getInstance();
        expect(helpScreen.state.isFaq).toEqual(true);
        helpScreen.toogleHelp();
        expect(helpScreen.state.isFaq).toEqual(false);
        helpScreen.toogleHelp();
        expect(helpScreen.state.isFaq).toEqual(true);
    });

    test('function goToTraidingApp mockLinking', () => {
        mockLinking();
        let helpScreen = renderer.create(<HelpScreen language={"ru"} navigation={{ setParams: () => { } }} />).getInstance();
        expect(helpScreen.goToTraidingApp()).toEqual();
    });    
})