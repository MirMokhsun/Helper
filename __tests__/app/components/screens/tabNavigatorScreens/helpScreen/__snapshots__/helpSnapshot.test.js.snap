// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help create snapshot HelpScreen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5F5F7",
      "flex": 1,
    }
  }
  testID="welcome"
>
  <View
    style={
      Object {
        "paddingHorizontal": 16,
        "paddingVertical": 8,
        "width": "100%",
      }
    }
  >
    <TouchableWithoutFeedback
      accessibilityLabel="HelpScreenBtnID"
      onPress={[Function]}
      testID="HelpScreenBtnID"
    >
      <View
        style={
          Object {
            "backgroundColor": "#EBECEE",
            "borderRadius": 16,
            "height": 32,
            "padding": 4,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "height": "100%",
              "left": 4,
              "position": "absolute",
              "top": 4,
              "width": "100%",
              "zIndex": 2,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "flex": 1,
                  "fontSize": 16,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                Object {
                  "color": "#000756",
                  "fontFamily": "Roboto-Bold",
                },
              ]
            }
          >
            FAQ
          </Text>
          <Text
            style={
              Array [
                Object {
                  "flex": 1,
                  "fontSize": 16,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                Object {
                  "color": "#9FA0B2",
                  "fontFamily": "Roboto-Light",
                },
              ]
            }
          >
            Словарь
          </Text>
        </View>
        <AnimatedComponent
          style={
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 12,
                "height": 24,
                "width": "50%",
              },
              Object {
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              },
            ]
          }
        />
      </View>
    </TouchableWithoutFeedback>
  </View>
  <FlatList
    data={
      Array [
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_0"
          onPress={[Function]}
          text="Как торговать?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_1"
          onPress={[Function]}
          text="Как внести депозит?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_2"
          onPress={[Function]}
          text="Stop Loss?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_3"
          onPress={[Function]}
          text="Что такое Forex?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_4"
          onPress={[Function]}
          text="Валюты на рынке Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_5"
          onPress={[Function]}
          text="География Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_6"
          onPress={[Function]}
          text="Как зарабатывать на Forex?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_7"
          onPress={[Function]}
          text="Как минимизировать риски?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_8"
          onPress={[Function]}
          text="Как выйти на Forex?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_9"
          onPress={[Function]}
          text="Forex и валюты: как работают валюты на рынке"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_10"
          onPress={[Function]}
          text="Объемы торговли на рынке Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_11"
          onPress={[Function]}
          text="Динамика курсов валют на рынке Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_12"
          onPress={[Function]}
          text="Ведет ли Forex к успеху?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_13"
          onPress={[Function]}
          text="Поведение валют на рынке Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_14"
          onPress={[Function]}
          text="Валюты на рынке Forex?"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_15"
          onPress={[Function]}
          text="Рынок Forex: история"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_16"
          onPress={[Function]}
          text="Джордж Сорос и его уникальная сделка"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_17"
          onPress={[Function]}
          text="Уоррен Баффет: философия успеха"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_18"
          onPress={[Function]}
          text="Первые шаги трейдера"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_19"
          onPress={[Function]}
          text="Forex депозит"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_20"
          onPress={[Function]}
          text="Пример заработка на Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_21"
          onPress={[Function]}
          text="Вывод прибыли"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_22"
          onPress={[Function]}
          text="Основы работы на рынке Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_23"
          onPress={[Function]}
          text="Стили торговли на Forex"
        />,
        <ListPicker
          currentTextAlign="left"
          id="DetailArticleScreen_24"
          onPress={[Function]}
          text="Как выбрать стратегию?"
        />,
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
